scripts/generate_service.go

```diff
@@ -0,0 +1,305 @@
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"go/format"
+	"log"
+	"os"
+	"strings"
+	"text/template"
+)
+
+type ComputeApiGenVersion int
+
+const (
+	none ComputeApiGenVersion = iota
+	v1
+	v0beta
+)
+
+var OrderedComputeApiVersions = []ComputeApiGenVersion{
+	v0beta,
+	v1,
+}
+
+type VersionInfo struct {
+	ClientNameSuffix string
+	ServiceName      string
+}
+
+type NormalisedData struct {
+	Type          string
+	LowestVersion ComputeApiGenVersion
+	UpdateVersion ComputeApiGenVersion
+	ExtraParams   []string
+}
+
+type TemplateData struct {
+	// Name for the resource in the new service API
+	ResourceName string
+
+	// Type of the resource object in the client. Generally but not always the ResourceName.
+	ClientType string
+
+	// Name of the Service of the client. Generally but not always the ResourceName pluralised.
+	ServiceType string
+
+	// Versions supported by this resource.
+	Versions []VersionInfo
+
+	// Versions supported by this resource for update.
+	UpdateVersions []VersionInfo
+
+	// Parameters to add to signatures/calls denormalised.
+	ExtraParamsSignature string
+	ExtraParamsCall      string
+}
+
+func main() {
+	clientTypeFlag := flag.String("type", "", "resource to generate a service for.")
+	scopeFlag := flag.String("scope", "", "scope if non-global. one of `region` or `zone`.")
+	parentFlag := flag.String("parent", "", "name of parent resource, if any.")
+	lowestVersionFlag := flag.String("lowestversion", "v1", "lowest version of the resource. one of `v1` or `v0beta`. defaults to v1.")
+	updateVersionFlag := flag.String("updateversion", "", "version at which update is supported. one of `v1` or `v0beta`.")
+	flag.Parse()
+
+	if *clientTypeFlag == "" {
+		flag.PrintDefaults()
+		log.Fatal("usage: go run generate_service.go -type $TYPE -scope $SCOPE -parent $PARENT -lowestversion $VERSION -updateversion $VERSION")
```

Comment:
`flag.usage()` prints some weird output about the binary for me; I prefer the message from `PrintDefaults` w/o that.